Misleading missing getStaticPaths error with configured redirects and dynamic routes

### Astro Info

```block
Astro                    v4.15.8
Node                     v18.20.3
System                   Linux (x64)
Package Manager          unknown
Output                   static
Adapter                  none
Integrations             none
```


### If this issue only occurs in one browser, which browser is a problem?

_No response_

### Describe the Bug

If I have a route like this:
```
/categories/[category]/
+-- index.astro
+-- [page].astro
```

In which each endpoint exports getStaticPaths, one for matching the base route with the category param, another for the category param with pagination.

If I include the following in `astro.config.mjs`:
```js
// ...
redirects: {
  '/categories/[category]': '/category/[category]/1',
},
// ...
```

Running `astro build` yields this error:
```
[ERROR] [build] Failed to call getStaticPaths for /categories/[category]
[GetStaticPathsRequired] `getStaticPaths()` function is required for dynamic routes. Make sure that you `export` a `getStaticPaths` function from your dynamic route.
```

### What's the expected result?

I would expect this setup to build without errors. It's my understanding that configured redirects are supported in static render mode, and even if this isn't the case, the error doesn't make sense and took far too long to for me to track down the source of the problem.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/withastro-astro-vyvufy?file=src%2Fpages%2Findex.astro

### Participation

- [ ] I am willing to submit a pull request for this issue.

--- a/.changeset/better-redirect-error.md
+++ b/.changeset/better-redirect-error.md
@@ -1,0 +1,7 @@
+---
+'astro': patch
+---
+
+Improves error message when a dynamic redirect destination does not match any existing route.
+
+Previously, configuring a redirect like `/categories/[category]` → `/categories/[category]/1` in static output mode would fail with a misleading "getStaticPaths required" error. Now, Astro detects this early and provides a clear error explaining that the destination does not match any existing route.

--- a/packages/astro/src/core/errors/errors-data.ts
+++ b/packages/astro/src/core/errors/errors-data.ts
@@ -1004,6 +1004,23 @@ export const UnsupportedExternalRedirect = {
 /**
  * @docs
  * @see
+ * - [Configured redirects](https://docs.astro.build/en/guides/routing/#configured-redirects)
+ * @description
+ * A dynamic redirect destination must match an existing route pattern.
+ * This error occurs when a redirect with dynamic parameters points to a destination
+ * that doesn't correspond to any page in your project.
+ */
+export const InvalidRedirectDestination = {
+	name: 'InvalidRedirectDestination',
+	title: 'Invalid redirect destination.',
+	message: (from: string, to: string) =>
+		`The redirect from "${from}" to "${to}" is invalid. The destination "${to}" does not match any existing route in your project.`,
+	hint: 'If you are redirecting to a specific page of a dynamic route (e.g., "/posts/[slug]/1"), this is not supported. The destination must be either a static path or a route pattern that matches an existing page (e.g., "/posts/[slug]/[page]").',
+} satisfies ErrorData;
+
+/**
+ * @docs
+ * @see
  * - [Dynamic routes](https://docs.astro.build/en/guides/routing/#dynamic-routes)
  * @description
  * A dynamic route param is invalid. This is often caused by an `undefined` parameter or a missing [rest parameter](https://docs.astro.build/en/guides/routing/#rest-parameters).

--- a/packages/astro/src/core/routing/manifest/create.ts
+++ b/packages/astro/src/core/routing/manifest/create.ts
@@ -13,6 +13,7 @@ import type { RouteData, RoutePart } from '../../../types/public/internal.js';
 import { toRoutingStrategy } from '../../app/index.js';
 import { SUPPORTED_MARKDOWN_FILE_EXTENSIONS } from '../../constants.js';
 import {
+	InvalidRedirectDestination,
 	MissingIndexForInternationalization,
 	UnsupportedExternalRedirect,
 } from '../../errors/errors-data.js';
@@ -360,6 +361,23 @@ function createRedirectRoutes(
 			});
 		}
 
+		const redirectRoute = routeMap.get(destination);
+
+		// If the source has dynamic params and the redirect will be prerendered,
+		// we need a valid redirectRoute to map them. Without it, the build will fail
+		// later with a misleading error. Catch this early and provide a clear error message.
+		if (
+			params.length > 0 &&
+			!redirectRoute &&
+			!URL.canParse(destination) &&
+			getPrerenderDefault(config)
+		) {
+			throw new AstroError({
+				...InvalidRedirectDestination,
+				message: InvalidRedirectDestination.message(from, destination),
+			});
+		}
+
 		routes.push({
 			type: 'redirect',
 			// For backwards compatibility, a redirect is never considered an index route.
@@ -372,7 +390,7 @@ function createRedirectRoutes(
 			pathname: pathname || void 0,
 			prerender: getPrerenderDefault(config),
 			redirect: to,
-			redirectRoute: routeMap.get(destination),
+			redirectRoute,
 			fallbackRoutes: [],
 			distURL: [],
 			origin: 'project',

--- a/packages/astro/test/fixtures/redirects/src/pages/categories/[category]/[page].astro
+++ b/packages/astro/test/fixtures/redirects/src/pages/categories/[category]/[page].astro
@@ -1,0 +1,20 @@
+---
+export function getStaticPaths() {
+	return [
+		{ params: { category: 'food', page: '1' } },
+		{ params: { category: 'food', page: '2' } },
+		{ params: { category: 'technology', page: '1' } },
+	];
+}
+
+const { category, page } = Astro.params;
+---
+<html>
+	<head>
+		<title>{category} - Page {page}</title>
+	</head>
+	<body>
+		<h1>{category}</h1>
+		<p>Page {page}</p>
+	</body>
+</html>

--- a/packages/astro/test/redirects.test.js
+++ b/packages/astro/test/redirects.test.js
@@ -349,4 +349,43 @@ describe('Astro.redirect output: "static"', () => {
 			assert.equal(html.includes('url=https://example.com/about'), true);
 		});
 	});
+
+	describe('Invalid redirect configurations', () => {
+		it('throws an error when dynamic redirect destination does not match any route', async () => {
+			// This tests the issue from https://github.com/withastro/astro/issues/12036
+			// where a redirect like /categories/[category] -> /categories/[category]/1
+			// produces a misleading "getStaticPaths required" error instead of
+			// a clear error about the invalid redirect destination.
+			fixture = await loadFixture({
+				root: './fixtures/redirects/',
+				output: 'static',
+				redirects: {
+					// The destination mixes a dynamic param [category] with a static segment "1"
+					// but the actual route is /categories/[category]/[page].astro
+					// The routeMap won't find "/categories/[category]/1" as a key
+					'/categories/[category]': '/categories/[category]/1',
+				},
+			});
+
+			try {
+				await fixture.build();
+				assert.fail('Expected build to fail');
+			} catch (e) {
+				// Should NOT be the misleading getStaticPaths error
+				assert.ok(
+					!e.message.includes('getStaticPaths()'),
+					`Expected error to not mention getStaticPaths, got: ${e.message}`,
+				);
+				// Should be our new clear error message
+				assert.ok(
+					e.message.includes('does not match any existing route'),
+					`Expected error to mention route matching issue, got: ${e.message}`,
+				);
+				assert.ok(
+					e.name === 'InvalidRedirectDestination',
+					`Expected error name to be InvalidRedirectDestination, got: ${e.name}`,
+				);
+			}
+		});
+	});
 });

